contents/java/JavaReflections.md

```diff
@@ -0,0 +1,298 @@
+
+The second line just means that this annotation can be applied to methods only.
+
+And the more important part is how the subscriber registry finds all its subscribing methods. The first step is to
+register a class as an event handler and an example of the code is like so:
+
+```java
+
+public class Sheep extends Animal {
+
+    private static final EventsCenter EVENTS_BUS = EventsCenter.getInstance();
+
+    public final String name;
+
+    public Sheep(String nameOfSheep) {
+        name = nameOfSheep;
+        EVENTS_BUS.register(this);
+    }
+
+    @Subscribe
+    public void handleWeatherChangeEvent(WeatherChangeEvent event) {
+        if (event.weather == Weather.RAIN) {
+            hide();
+        }
+    }
+
+    // Other methods here...
+    
+}
+```
+
+An example implementation of the `EventsCenter` (with a lot of details left out for simplicity) is like so:
+
+```java
+
+public void register(Class<?> clazz) {
+    findAllEventHandlersInClass(clazz);
+}
+
+public void findAllEventHandlersInClass(Class<?> clazz) {
+    // TypeToken class is provided by Google Guava reflection library
+    Set<? extends Class<?>> supertypes = TypeToken.of(clazz).getTypes().rawTypes();
+
+    for (Class<?> supertype : supertypes) {
+          for (Method method : supertype.getDeclaredMethods()) {
+            if (method.isAnnotationPresent(Subscribe.class) {
+                registerSubscriber(method);
+            }
+        }
+    }
+}
+
+```
+
+The first line of the `findAllEventHandlersInClass` method finds all classes and its parent classes of the registered
+class and converts it to a set. That is if you registered `Sheep extends Animal` as an event handler to the method,
+both `Sheep` and `Animal` will be captured by the first line. The following lines will then examine all their methods
+(during runtime!) for the `Subscribe` annotation and register the method so that it will receive the specified event
+when it is fired. Of course this implementation leaves out a lot of details but you get the idea of how java reflections
+works.
+
+### Disadvantages of reflections
+
+While Java reflections are powerful, you should not immediately jump on the reflections ship. This is because there are
+some drawbacks whenever you use reflections in your project. The following are some points you should consider before
+using Java reflections:
+
+* Reflections convert a compile-time error to a potentially destructive run-time error.
+
+Compile time errors are easy to catch. Whenever you compile your code, the compiler cleverly spots any error you
+missed and points it out (along with line number and other useful information) to you before quitting. But by using
+reflections, you are bypassing these checks because there is no way to check such errors during compile time. These 
+uncaught errors may cause your program to fail during runtime instead, turning into runtime errors.
```

Comment:
Indent this paragraph by 2 spaces if you want it to appear as part of the previous bullet point.